### ISSUE TITLE
Bugs+semantics of regex and channel matching

### ISSUE THREAD

**Issue by vasslitvinov** [chapel-lang/chapel#20864]
Bug 1: when given the regex `^`, iter `channel.matches()` matches before each character in the channel and also after the last character in the channel. The desired behavior: for this regex it should match only at the start of the channel.

Bug 2: when given the regex `$`, proc `regex.search()` produces a match with byteOffset = 0. The desired behavior: byteOffset should be the size of the string, i.e., it should point at the position after the last character in the string.

**Comment by vasslitvinov** [comment]
@mppf and everyone—channel positions look to be advanced incorrectly in the following cases. How do we want them to work?

Case 1: invoking iter `channel.matches()` or proc `channel.search()` with the regex `^$`. I expect the channel position to be unchanged, whereas currently either function advances it by 1. Or perhaps the channel position should be at the end of the channel, because this is what the documentation states.

Case 1x: where should I find the channel's position after `channel.matches()` or `channel.search()` with the regex `^` when the position is already not at the start of the channel?

Case 2: should iter `channel.matches()` advance the channel the same way whether or not proc `channel.extractMatch()` is invoked in the loop body? Currently when extractMatch() is not invoked it does not advance the channel and loops infinitely.

**Comment by vasslitvinov** [comment]
Another similar bug is in #20431.

**Comment by vasslitvinov** [comment]
For the record, I announced this issue to the group on 10/21 and got no objections or other feedback.

**Comment by bradcray** [comment]
Given the timing and other things going on that week, I'd try again, particularly from those who are re2- and IO-knowledgeable like @mppf.

**Comment by vasslitvinov** [comment]
Case 3: what should be the position of the channel after `r.search(compile("^\\d+"))` when there is no match and the channel is (a) empty or (b) not empty? The current behavior is reflected in `test/regexp/ferguson/rechan.chpl`

**Comment by vasslitvinov** [comment]
The current behavior of `proc _channel.search` leaves the channel position at the start of a match if there is one. However this is counter-intuitive to me. Since we have established a match, we have looked at all the contents of the channel through the end of the match. So that's where I expect the ending position to be. From another angle...

If I invoke search() on a channel repeatedly, I expect the subsequent search to start where the previous search finished. At least this is the behavior of the incremental search in emacs. :)

What is the rationale for the current behavior? BTW the chpldoc needs to be clearer about it.
